File: mcp/linktor-mcp-server/src/tools/vre.ts
// ============================================
// VRE (Visual Response Engine) Tools
// ============================================

import type { Tool } from '@modelcontextprotocol/sdk/types.js';
import type { LinktorClient } from '../api/client.js';

export const vreToolDefinitions: Tool[] = [
  {
    name: 'render_template',
    description: 'Render a VRE template to an image. Returns base64-encoded image data that can be sent to messaging channels.',
    inputSchema: {
      type: 'object',
      properties: {
        tenant_id: {
          type: 'string',
          description: 'The tenant ID',
        },
        template_id: {
          type: 'string',
          enum: ['menu_opcoes', 'card_produto', 'status_pedido', 'lista_produtos', 'confirmacao', 'cobranca_pix'],
          description: 'The VRE template to render',
        },
        data: {
          type: 'object',
          description: 'Template data (varies by template type)',
          additionalProperties: true,
        },
        channel: {
          type: 'string',
          enum: ['whatsapp', 'telegram', 'email', 'web'],
          description: 'Target channel for optimized rendering (default: whatsapp)',
          default: 'whatsapp',
        },
        format: {
          type: 'string',
          enum: ['webp', 'png', 'jpeg'],
          description: 'Output image format (default: webp)',
          default: 'webp',
        },
      },
      required: ['tenant_id', 'template_id', 'data'],
    },
  },
  {
    name: 'render_and_send',
    description: 'Render a VRE template and send it directly to a conversation. Combines rendering and sending in one operation.',
    inputSchema: {
      type: 'object',
      properties: {
        conversation_id: {
          type: 'string',
          description: 'The conversation ID to send the rendered image to',
        },
        template_id: {
          type: 'string',
          enum: ['menu_opcoes', 'card_produto', 'status_pedido', 'lista_produtos', 'confirmacao', 'cobranca_pix'],
          description: 'The VRE template to render',
        },
        data: {
          type: 'object',
          description: 'Template data (varies by template type)',
          additionalProperties: true,
        },
        caption: {
          type: 'string',
          description: 'Optional custom caption for the image (auto-generated if not provided)',
        },
        follow_up_text: {
          type: 'string',
          description: 'Optional text message to send after the image',
        },
      },
      required: ['conversation_id', 'template_id', 'data'],
    },
  },
  {
    name: 'list_templates',
    description: 'List available VRE templates with their schemas and example data.',
    inputSchema: {
      type: 'object',
      properties: {
        tenant_id: {
          type: 'string',
          description: 'Optional tenant ID to include custom templates',
        },
      },
    },
  },
  {
    name: 'preview_template',
    description: 'Preview a VRE template with sample data. Returns the rendered image for inspection.',
    inputSchema: {
      type: 'object',
      properties: {
        template_id: {
          type: 'string',
          enum: ['menu_opcoes', 'card_produto', 'status_pedido', 'lista_produtos', 'confirmacao', 'cobranca_pix'],
          description: 'The VRE template to preview',
        },
        data: {
          type: 'object',
          description: 'Optional custom data for preview (uses default sample data if not provided)',
          additionalProperties: true,
        },
      },
      required: ['template_id'],
    },
  },
  // Visual response tools for bot integration
  {
    name: 'mostrar_menu',
    description: 'Display a visual menu with numbered options for the user to choose from. Use this when you need to present multiple options clearly. Returns an image with caption.',
    inputSchema: {
      type: 'object',
      properties: {
        titulo: {
          type: 'string',
          description: 'Menu title',
        },
        opcoes: {
          type: 'array',
          description: 'List of options (max 8)',
          items: {
            type: 'object',
            properties: {
              label: { type: 'string', description: 'Option text' },
              icone: { type: 'string', description: 'Icon name (pedido, catalogo, entrega, financeiro, atendente, etc)' },
              descricao: { type: 'string', description: 'Optional description' },
            },
            required: ['label'],
          },
          maxItems: 8,
        },
        conversation_id: {
          type: 'string',
          description: 'Conversation ID to send the menu to',
        },
      },
      required: ['titulo', 'opcoes', 'conversation_id'],
    },
  },
  {
    name: 'mostrar_card_produto',
    description: 'Display a visual product card with image, price, and stock information. Use for showcasing a single product.',
    inputSchema: {
      type: 'object',
      properties: {
        nome: { type: 'string', description: 'Product name' },
        preco: { type: 'number', description: 'Product price' },
        unidade: { type: 'string', description: 'Unit (kg, un, cx, etc)', default: 'un' },
        imagem_url: { type: 'string', description: 'Product image URL' },
        estoque: { type: 'string', enum: ['disponivel', 'baixo', 'indisponivel'], description: 'Stock status' },
        descricao: { type: 'string', description: 'Product description' },
        conversation_id: { type: 'string', description: 'Conversation ID' },
      },
      required: ['nome', 'preco', 'conversation_id'],
    },
  },
  {
    name: 'mostrar_status_pedido',
    description: 'Display a visual order status with progress timeline. Use to show order tracking information.',
    inputSchema: {
      type: 'object',
      properties: {
        numero_pedido: { type: 'string', description: 'Order number' },
        status_atual: {
          type: 'string',
          enum: ['recebido', 'separacao', 'faturado', 'transporte', 'entregue'],
          description: 'Current order status'
        },
        itens_resumo: { type: 'string', description: 'Items summary' },
        valor_total: { type: 'number', description: 'Total order value' },
        previsao_entrega: { type: 'string', description: 'Delivery estimate' },
        motorista: { type: 'string', description: 'Driver name' },
        conversation_id: { type: 'string', description: 'Conversation ID' },
      },
      required: ['numero_pedido', 'status_atual', 'conversation_id'],
    },
  },
  {
    name: 'mostrar_lista_produtos',
    description: 'Display a visual product list for comparison. Use when showing multiple products side by side.',
    inputSchema: {
      type: 'object',
      properties: {
        titulo: { type: 'string', description: 'List title' },
        produtos: {
          type: 'array',
          description: 'List of products (max 6)',
          items: {
            type: 'object',
            properties: {
              nome: { type: 'string' },
              preco: { type: 'number' },
              unidade: { type: 'string' },
              estoque: { type: 'string' },
            },
            required: ['nome', 'preco'],
          },
          maxItems: 6,
        },
        conversation_id: { type: 'string', description: 'Conversation ID' },
      },
      required: ['titulo', 'produtos', 'conversation_id'],
    },
  },
  {
    name: 'mostrar_confirmacao',
    description: 'Display a visual order/action confirmation with item summary and total. Use for final confirmation before processing.',
    inputSchema: {
      type: 'object',
      properties: {
        titulo: { type: 'string', description: 'Confirmation title' },
        itens: {
          type: 'array',
          description: 'List of items',
          items: {
            type: 'object',
            properties: {
              nome: { type: 'string' },
              quantidade: { type: 'string' },
              preco: { type: 'number' },
            },
          },
        },
        valor_total: { type: 'number', description: 'Total value' },
        previsao_entrega: { type: 'string', description: 'Delivery estimate' },
        conversation_id: { type: 'string', description: 'Conversation ID' },
      },
      required: ['titulo', 'valor_total', 'conversation_id'],
    },
  },
  {
    name: 'mostrar_cobranca_pix',
    description: 'Display a PIX payment QR code with copy-paste code. Use for payment collection.',
    inputSchema: {
      type: 'object',
      properties: {
        valor: { type: 'number', description: 'Payment amount' },
        pix_payload: { type: 'string', description: 'PIX copy-paste code' },
        qr_code_url: { type: 'string', description: 'QR code image URL' },
        numero_pedido: { type: 'string', description: 'Order number' },
        expiracao: { type: 'string', description: 'Expiration time' },
        conversation_id: { type: 'string', description: 'Conversation ID' },
      },
      required: ['valor', 'pix_payload', 'conversation_id'],
    },
  },
];

export function registerVRETools(
  handlers: Map<string, (args: Record<string, unknown>) => Promise<unknown>>,
  client: LinktorClient
): void {
  handlers.set('render_template', async (args) => {
    return client.vre.render({
      tenant_id: args.tenant_id as string,
      template_id: args.template_id as string,
      data: args.data as Record<string, unknown>,
      channel: args.channel as string | undefined,
      format: args.format as string | undefined,
    });
  });

  handlers.set('render_and_send', async (args) => {
    return client.vre.renderAndSend({
      conversation_id: args.conversation_id as string,
      template_id: args.template_id as string,
      data: args.data as Record<string, unknown>,
      caption: args.caption as string | undefined,
      follow_up_text: args.follow_up_text as string | undefined,
    });
  });

  handlers.set('list_templates', async (args) => {
    return client.vre.listTemplates(args.tenant_id as string | undefined);
  });

  handlers.set('preview_template', async (args) => {
    return client.vre.preview({
      template_id: args.template_id as string,
      data: args.data as Record<string, unknown> | undefined,
    });
  });

  // Visual bot tools (render and send in one step)
  const visualTools = [
    'mostrar_menu',
    'mostrar_card_produto',
    'mostrar_status_pedido',
    'mostrar_lista_produtos',
    'mostrar_confirmacao',
    'mostrar_cobranca_pix',
  ];

  for (const toolName of visualTools) {
    handlers.set(toolName, async (args) => {
      const conversationId = args.conversation_id as string;
      const templateId = toolName.replace('mostrar_', '').replace('card_', '').replace('lista_', '');

      // Map tool name to template ID
      const templateMap: Record<string, string> = {
        'mostrar_menu': 'menu_opcoes',
        'mostrar_card_produto': 'card_produto',
        'mostrar_status_pedido': 'status_pedido',
        'mostrar_lista_produtos': 'lista_produtos',
        'mostrar_confirmacao': 'confirmacao',
        'mostrar_cobranca_pix': 'cobranca_pix',
      };

      // Remove conversation_id from data
      const { conversation_id, ...data } = args;

      return client.vre.renderAndSend({
        conversation_id: conversationId,
        template_id: templateMap[toolName] || templateId,
        data: data as Record<string, unknown>,
      });
    });
  }
}

File: sdks/java/src/main/java/io/linktor/types/VRE.java
package io.linktor.types;

import com.google.gson.annotations.SerializedName;

import java.util.Map;

/**
 * VRE (Visual Response Engine) types
 */
public class VRE {

    /**
     * Output format for rendered images
     */
    public enum OutputFormat {
        @SerializedName("png") PNG,
        @SerializedName("webp") WEBP,
        @SerializedName("jpeg") JPEG
    }

    /**
     * Channel type for VRE rendering
     */
    public enum ChannelType {
        @SerializedName("whatsapp") WHATSAPP,
        @SerializedName("telegram") TELEGRAM,
        @SerializedName("web") WEB,
        @SerializedName("email") EMAIL
    }

    /**
     * Available template types
     */
    public enum TemplateType {
        @SerializedName("menu_opcoes") MENU_OPCOES,
        @SerializedName("card_produto") CARD_PRODUTO,
        @SerializedName("status_pedido") STATUS_PEDIDO,
        @SerializedName("lista_produtos") LISTA_PRODUTOS,
        @SerializedName("confirmacao") CONFIRMACAO,
        @SerializedName("cobranca_pix") COBRANCA_PIX
    }

    /**
     * Order status for status_pedido template
     */
    public enum OrderStatus {
        @SerializedName("recebido") RECEBIDO,
        @SerializedName("separacao") SEPARACAO,
        @SerializedName("faturado") FATURADO,
        @SerializedName("transporte") TRANSPORTE,
        @SerializedName("entregue") ENTREGUE
    }

    /**
     * Render request
     */
    public static class RenderRequest {
        @SerializedName("tenant_id")
        private String tenantId;

        @SerializedName("template_id")
        private String templateId;

        private Map<String, Object> data;

        private String channel;

        private String format;

        private Integer width;

        private Integer quality;

        private Double scale;

        // Getters
        public String getTenantId() { return tenantId; }
        public String getTemplateId() { return templateId; }
        public Map<String, Object> getData() { return data; }
        public String getChannel() { return channel; }
        public String getFormat() { return format; }
        public Integer getWidth() { return width; }
        public Integer getQuality() { return quality; }
        public Double getScale() { return scale; }

        // Builder
        public static Builder builder() { return new Builder(); }

        public static class Builder {
            private final RenderRequest request = new RenderRequest();

            public Builder tenantId(String tenantId) {
                request.tenantId = tenantId;
                return this;
            }

            public Builder templateId(String templateId) {
                request.templateId = templateId;
                return this;
            }

            public Builder templateId(TemplateType templateId) {
                request.templateId = templateId.name().toLowerCase();
                return this;
            }

            public Builder data(Map<String, Object> data) {
                request.data = data;
                return this;
            }

            public Builder channel(String channel) {
                request.channel = channel;
                return this;
            }

            public Builder channel(ChannelType channel) {
                request.channel = channel.name().toLowerCase();
                return this;
            }

            public Builder format(String format) {
                request.format = format;
                return this;
            }

            public Builder format(OutputFormat format) {
                request.format = format.name().toLowerCase();
                return this;
            }

            public Builder width(Integer width) {
                request.width = width;
                return this;
            }

            public Builder quality(Integer quality) {
                request.quality = quality;
                return this;
            }

            public Builder scale(Double scale) {
                request.scale = scale;
                return this;
            }

            public RenderRequest build() { return request; }
        }
    }

    /**
     * Render response
     */
    public static class RenderResponse {
        @SerializedName("image_base64")
        private String imageBase64;

        private String caption;

        private int width;

        private int height;

        private String format;

        @SerializedName("render_time_ms")
        private int renderTimeMs;

        @SerializedName("size_bytes")
        private Long sizeBytes;

        @SerializedName("cache_hit")
        private Boolean cacheHit;

        // Getters
        public String getImageBase64() { return imageBase64; }
        public String getCaption() { return caption; }
        public int getWidth() { return width; }
        public int getHeight() { return height; }
        public String getFormat() { return format; }
        public int getRenderTimeMs() { return renderTimeMs; }
        public Long getSizeBytes() { return sizeBytes; }
        public Boolean getCacheHit() { return cacheHit; }
    }

    /**
     * Render and send request
     */
    public static class RenderAndSendRequest {
        @SerializedName("conversation_id")
        private String conversationId;

        @SerializedName("template_id")
        private String templateId;

        private Map<String, Object> data;

        private String caption;

        @SerializedName("follow_up_text")
        private String followUpText;

        // Getters
        public String getConversationId() { return conversationId; }
        public String getTemplateId() { return templateId; }
        public Map<String, Object> getData() { return data; }
        public String getCaption() { return caption; }
        public String getFollowUpText() { return followUpText; }

        // Builder
        public static Builder builder() { return new Builder(); }

        public static class Builder {
            private final RenderAndSendRequest request = new RenderAndSendRequest();

            public Builder conversationId(String conversationId) {
                request.conversationId = conversationId;
                return this;
            }

            public Builder templateId(String templateId) {
                request.templateId = templateId;
                return this;
            }

            public Builder templateId(TemplateType templateId) {
                request.templateId = templateId.name().toLowerCase();
                return this;
            }

            public Builder data(Map<String, Object> data) {
                request.data = data;
                return this;
            }

            public Builder caption(String caption) {
                request.caption = caption;
                return this;
            }

            public Builder followUpText(String followUpText) {
                request.followUpText = followUpText;
                return this;
            }

            public RenderAndSendRequest build() { return request; }
        }
    }

    /**
     * Render and send response
     */
    public static class RenderAndSendResponse {
        @SerializedName("message_id")
        private String messageId;

        @SerializedName("image_url")
        private String imageUrl;

        private String caption;

        // Getters
        public String getMessageId() { return messageId; }
        public String getImageUrl() { return imageUrl; }
        public String getCaption() { return caption; }
    }

    /**
     * Template definition
     */
    public static class Template {
        private String id;
        private String name;
        private String description;
        private Map<String, Object> schema;

        // Getters
        public String getId() { return id; }
        public String getName() { return name; }
        public String getDescription() { return description; }
        public Map<String, Object> getSchema() { return schema; }
    }

    /**
     * List templates response
     */
    public static class ListTemplatesResponse {
        private java.util.List<Template> templates;

        public java.util.List<Template> getTemplates() { return templates; }
    }

    /**
     * Preview request
     */
    public static class PreviewRequest {
        private Map<String, Object> data;

        public PreviewRequest() {}

        public PreviewRequest(Map<String, Object> data) {
            this.data = data;
        }

        public Map<String, Object> getData() { return data; }
        public void setData(Map<String, Object> data) { this.data = data; }
    }

    /**
     * Preview response
     */
    public static class PreviewResponse {
        @SerializedName("image_base64")
        private String imageBase64;

        private int width;
        private int height;

        // Getters
        public String getImageBase64() { return imageBase64; }
        public int getWidth() { return width; }
        public int getHeight() { return height; }
    }
}
